### PR TITLE
<memory>: Fix atomic smart pointers array type interaction

### LINKED ISSUE #1332: <memory>: atomic smart pointers do not work with array elements
**Describe the bug**
`atomic<shared_ptr<Ty>>` and `atomic<weak_ptr<Ty>>` does not work when `Ty` is and an array `{ Ty = T[] || T[N] }`. 
No sure if this is intended. [[util.smartptr.atomic]](http://eel.is/c++draft/util.smartptr.atomic) does not really specify.

**Code test case**
[Godbolt example](https://godbolt.org/z/5G7MfT) (Uncomment lines)
```C++
#include <memory>

int main() {
    using namespace std;
    shared_ptr<int> sptr0;
    shared_ptr<int[]> sptr1;
    shared_ptr<int[2]> sptr2;
    
    atomic<shared_ptr<int>> atomic_sptr0;
    atomic<shared_ptr<int[]>> atomic_sptr1;
    atomic<shared_ptr<int[2]>> atomic_sptr2;

    atomic_sptr0 = sptr0;
    atomic_sptr1 = sptr1;                                      // <-- Broken ?
    atomic_sptr2 = sptr2;                                      // <-- Broken ?
    
    weak_ptr<int> wptr0;
    weak_ptr<int[]> wptr1;
    weak_ptr<int[2]> wptr2;
    
    atomic<weak_ptr<int>> atomic_wptr0;
    atomic<weak_ptr<int[]>> atomic_wptr1;
    atomic<weak_ptr<int[2]>> atomic_wptr2;
    
    atomic_wptr0 = wptr0;
    atomic_wptr1 = wptr1;                                      // <-- Broken ?
    atomic_wptr2 = wptr2;                                      // <-- Broken ?

    return 0;
}
```

**Expected behavior**
For the code to be compiled.

**STL version**
All of them but locally I ran into the problem with:
    ```
    Microsoft Visual Studio Community 2019 Preview
    Version 16.8.0 Preview 3.2
    ```
**Additional context**
The issue seems come from a mismatch between the type used to represent the internal pointer in  `shared_ptr ` and  `weak_ptr ` compared to  `atomic<x_ptr<>> `.  `shared_ptr<T>::element_type` and  `weak_ptr<T>::element_type` is of type `remove_extent_t<T>` while `atomic<x_ptr<T>> `  internal representation is just `T*`. So it ends up trying to perform operations between `T[] *` and `T*`

I implemented my own fix (0fd8850c0cd9adb34c6b9a63e7adf2b513f7d082) so I could continue with my projects. 
Simply changed:
```diff
- _Ty*
+ remove_extent_t<_Ty>*
```
Also extended the one test suit (975ac24831d73ed3f736df8d47f5b1e2e3f6a9f7) as a sanity check. 

(Does not test `wait`, `notify_one` and `notify_all`)

### PR DESCRIPTION
**Pull request by xmas92**
Fixes #1332

*Notes*
Have not tested if it affects `wait`, `notify_one` and `notify_all`. Not that it should, but I am not well versed in the smart pointer implementations.

### CODE REVIEW

File: tests/std/tests/P0718R2_atomic_smart_ptrs/test.cpp [hunk 1]
```diff
@@ -183,6 +200,326 @@ void test_weak_ptr_compare_exchange_strong() {
     }
 }
 
+// Repeat test for unbounded array type.
+
```
Comment (miscco) [inline tests/std/tests/P0718R2_atomic_smart_ptrs/test.cpp:204]:
```suggestion
```

File: tests/std/tests/P0718R2_atomic_smart_ptrs/test.cpp [hunk 2]
```diff
@@ -183,6 +200,326 @@ void test_weak_ptr_compare_exchange_strong() {
...
+            assert(weak_ptr_equal(local, wp1));
+        } else {
+            assert(weak_ptr_equal(local, wp0));
+        }
+        this_thread::yield();
+    }
+}
+
+// Repeat test for bounded array type.
+
```
Comment (miscco) [inline tests/std/tests/P0718R2_atomic_smart_ptrs/test.cpp:364]:
```suggestion
```

**Comment by StephanTLavavej** [comment]
Thanks! Should there be test coverage for multidimensional arrays? (This is the first thing I think of when I see `remove_extent_t`.) It doesn't have to be as thorough as the single-dimensional array coverage, just enough to indicate that this is the correct type trait to use.

**Comment by xmas92** [comment]
Added a compile test for multidimensional arrays (member and non-member functions). If you also want to test the functionality I think the best thing is to just repeat the original test for the different types. Could refactor it into template functions.

As originally noted the test do not touch the `wait/notify` mechanism. Not even if it compiles. 
Is there some other test that test these and could be used to check that it still works for array types. 

Otherwise I could not think of a simple way to test whether it compiles except writing some producer/consumer type of functions which use `wait/notify`.

**Comment by AlexGuteniev** [comment]
STL/tests/std/tests/P1135R6_atomic_wait/test.cpp is where atomic wait is tested

**Review by strega-nil-ms — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]
Notes:

This in theory changes ABI of `_Atomic_ptr_base`; however, it only changes name manglings of non-public names, and changes the members of `_Atomic_ptr_base<T[]>` from:

```cxx
{
    atomic<T(*)[]> _Ptr;
    _Locked_pointer<_Ref_count_base> _Repptr;
}
```

to

```cxx
{
    atomic<T*> _Ptr;
    _Locked_pointer<_Ref_count_base> _Repptr;
}
```

which should not change ABI/layout at all.

**Review by strega-nil-ms — DISMISSED** [review]
`_Seed` should not be a template, since there's no reason for it to be a template. Additionally, an existing, correct, narrowing conversion now fails.

**Review by strega-nil-ms — APPROVED** [review]

File: tests/std/include/test_atomic_wait.hpp [hunk 3]
```diff
@@ -108,7 +108,8 @@ void test_notify_all_notifies_all(UnderlyingType old_value, const UnderlyingType
 template <class UnderlyingType>
 void test_notify_all_notifies_all_ptr(UnderlyingType old_value, const UnderlyingType new_value,
     const std::chrono::steady_clock::duration waiting_duration) {
-    test_notify_all_notifies_all_impl<std::atomic, UnderlyingType>(old_value, new_value, waiting_duration);
+    // increased waiting_duration because timing assumption might not hold for atomic smart pointers
+    test_notify_all_notifies_all_impl<std::atomic, UnderlyingType>(old_value, new_value, 3 * waiting_duration);
```
Comment (StephanTLavavej) [inline tests/std/include/test_atomic_wait.hpp:112]: I'm slightly concerned about this change because the test suite should never contain timing assumptions that could lead to sporadic failures. That said, no change requested, because this is clearly a pre-existing issue.

File: tests/std/include/test_atomic_wait.hpp [hunk 4]
```diff
@@ -248,6 +264,25 @@ inline void test_atomic_wait() {
     test_notify_all_notifies_all(three_chars{1, 1, 3}, three_chars{1, 2, 3}, waiting_duration);
     test_notify_all_notifies_all(big_char_like{'a'}, big_char_like{'b'}, waiting_duration);
 
+    test_notify_all_notifies_all_ptr(std::make_shared<int>('a'), std::make_shared<int>('a'), waiting_duration);
+    test_notify_all_notifies_all_ptr(
+        std::weak_ptr{std::make_shared<int>('a')}, std::weak_ptr{std::make_shared<int>('a')}, waiting_duration);
```
Comment (StephanTLavavej) [inline tests/std/include/test_atomic_wait.hpp:269]: No change requested: Nice catch, `test_notify_all_notifies_all_ptr()` was defined but never used before.

File: stl/inc/memory [hunk 5]
```diff
@@ -3872,7 +3873,7 @@ protected:
         _Ptr.notify_all();
     }
 
-    atomic<_Ty*> _Ptr{nullptr};
+    atomic<remove_extent_t<_Ty>*> _Ptr{nullptr};
```
Comment (StephanTLavavej) [inline stl/inc/memory:3876]: No change requested: Regarding the ABI implications that @strega-nil-ms mentioned in https://github.com/microsoft/STL/pull/1339#pullrequestreview-961181700 - while changing data members is always concerning from an ABI perspective, I agree that the actual bits being stored are the same. Additionally, the previous code would generally fail to compile for simple cases like `store()`, which indicates that we shouldn't have to worry about mismatch issues. (The one time we can make unrestricted changes to layout is when the code couldn't compile at all before - this is not quite "wouldn't compile at all" but close.)

File: tests/std/include/test_atomic_wait.hpp [hunk 6]
```diff
@@ -228,6 +229,21 @@ inline void test_atomic_wait() {
     test_atomic_wait_func_ptr(std::make_shared<int>('a'), std::make_shared<int>('a'), waiting_duration);
     test_atomic_wait_func_ptr(
         std::weak_ptr{std::make_shared<int>('a')}, std::weak_ptr{std::make_shared<int>('a')}, waiting_duration);
+    test_atomic_wait_func_ptr(std::make_shared<int[]>(0), std::make_shared<int[]>(0), waiting_duration);
```
Comment (StephanTLavavej) [inline tests/std/include/test_atomic_wait.hpp:232]: No change requested: I note that the zero-runtime-length arrays being used in the test cases are unusual; however, I believe this is conforming, and the code also tests one-element arrays, so I have no objection to this (even if I would have used non-zero numbers for all).

File: tests/std/tests/P0718R2_atomic_smart_ptrs/test.cpp [hunk 7]
```diff
@@ -222,6 +557,25 @@ void ensure_nonmember_calls_compile() {
...
+    loaded      = constInstance.load();
+    loaded      = instance;
+    loaded      = constInstance;
+    instance.store(loaded);
+    loaded = instance.exchange(instance);
+    loaded = instance.exchange(constInstance);
+    if (instance.compare_exchange_weak(loaded, constInstance)) {
+        // intentionally empty
+    }
+    if (instance.compare_exchange_strong(loaded, constInstance)) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0718R2_atomic_smart_ptrs/test.cpp:574]: No change requested: We conventionally add newlines between non-chained `if`-statements. However, this was copied from pre-existing code immediately above, and it's not worth resetting testing by pushing a commit.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this bug, and congratulations on your first microsoft/STL commit! :tada: :smile_cat: :rocket:

(And apologies for taking so long to review this! :scream_cat:)
